libcsf/man3/SetMV.3
.lf 1 SetMV.3
.\" WARNING! THIS FILE WAS GENERATED AUTOMATICALLY BY c2man!
.\" DO NOT EDIT! CHANGES MADE TO THIS FILE WILL BE LOST!
.TH "SetMV" 3 "13 August 1999" "c2man setmv.c"
.SH "NAME"
SetMV \- set a memory location to a missing value
.SH "SYNOPSIS"
.ft B
#include "csf.h"
.br
.sp
void SetMV
.br
(
.br
	const MAP *m,
.br
	void *c
.br
);
.ft R
.SH "PARAMETERS"
.TP
.B "const MAP *m"
map handle
.TP
.B "void *c"
write-only. location set to missing value
.SH "DESCRIPTION"
SetMV sets a memory location to a missing value
(using the application cell representation).
SetMV is quite slow but handy as in the example
below. In general one should use assignment for
integers (e.g. v = MV_UINT1) or the macro's
SET_MV_REAL4 and SET_MV_REAL8
.SH "EXAMPLE"
.lf 1 examples/border.tr
.DS
 
 #include <stdlib.h>
 #include "csf.h"
 
 /* set border to MV
  * with minimal checking
  */
 
 void main(int argc, char *argv[] )
 {
   MAP *map;                      
   void *mv;
   size_t i,nrRows,nrCols;
 
   if (argc != 2)
   {
    fprintf(stderr,"%s: no file specified\\n",argv[0]);
    exit(1);
   }
 
   map = Mopen(argv[1], M_READ_WRITE);
   RuseAs(map, CR_REAL8); 
 
   mv = Rmalloc(map, 1);
   SetMV(map, mv);
 
   nrRows = RgetNrRows(map);
   nrCols = RgetNrCols(map);
 
   for(i=0; i < RgetNrCols(map); i++) { 
    RputCell(map, 0       , i, mv);
    RputCell(map, nrRows-1, i, mv);
   }
   for(i=0; i < RgetNrRows(map); i++) { 
    RputCell(map, i, 0       , mv);
    RputCell(map, i, nrCols-1, mv);
   }
 
   free(mv);
   Mclose(map);
 
   exit(0);
 }
 
.DE
.lf 37 SetMV.3
.SH "SEE ALSO"
SetMVcellRepr(3)


src/CHANGES
/*
CHANGES log (summary)
============================
Changes written after a version number apply to the next the version.
date: version (program name)

dec 92, spil 2.06:
	added method ossfim, dump_cvv(); dynamic data maximum.
jan 93, spil 2.07:
	dynamic allocation data, unbounded maximum.
mar 93, spil 2.08:
	corrected block discretizing
apr 93, spil 2.09:
	added lex variogram expression parsing (rd_vario.[lc]) ## later removed
apr 93, spil 3.0:
	added full command file parser with lexical analyser
	added sample variogram calculation (sem)

jun 93, gstat 0.99
	spil 3.0 + sem -> gstat 0.99; "gstat [-]sem" ## later: -e semivariance

jun 93, gstat 0.99c:
	removed bug in local/ordinary kriging
jun 93, gstat 0.99d:
	coupled gstat with meschach library; removed numerical recip matrix code.
jun 93, gstat 0.99e:
	solved bug in semi().
aug 93, gstat 0.99f:
	added MapSign().
aug 93, gstat 0.99g:
	added time coordinate to data, distance(x,t) not solved.
	corrected distance bug when d->force in select.c
	added variogram->max_[range/sill] 
aug 93, gstat 0.99h
	added max_ldl structure, and correction with -L option
sep 93, gstat 0.99i
	dynamic allocation of nested variogram models (unbounded maximum)
sep 93, gstat 0.99j
	added universal kriging
oct 93, gstat 1.0  
	release without univ.kr., added variography check (ic,lmc).
nov 93, gstat 1.1  
	renewed error handling, corrected variography check
	corrected bug in memory allocation in ReadLegend();
	released gstat 1.1 on dec 1, 1993

dec 93, gstat 1.2 
	changed the VARIOGRAM structure, using V_model and V_model_type;
	[ check UK, consider REML? ]
	added 4x4 Gauss quadrature block descretizing (Carr & Palmer)
	removed the attr2 field from POINT, changed semi() accordingly
	removed bug in positioning results from ID, if n_vars > 1
	changed UK, added LM. def: with intercept, other variables with X
	revised sem.c , added sem_main.c, made CheckX();
apr 94, gstat 1.2a 
	removed bug in sem_main.c (log, and I for data2);
	removed bug in do_mask.c, warning for coordinates when PT_XY
	removed bug in direct.c (117), added ER_DOMAIN
apr 94, gstat 1.2b
	added gram_sch.c: Gram_Schmidt checking of matrices: option -g
	added # comment-style and flex -l compatability (GNU flex 2.4.6)
	removed bug in header cross semivariogram (pseudo <--> symmetric)
	changed Quant() and Quantile, so they equal Splus' quantile()
june 94, gstat 1.2c
	removed bug when using both Valid and XValid (reversed order)
	removed method tin, lss, ... added method xy
june 94, gstat 1.2d
	removed bug in Skrige() (Oops, I forgot to subtract sk_mean)
june 94, gstat 1.2e
	added reading `points(xxx):"map.csf"'; points from csf map.
    released 1.2f
june 94, gstat 1.2g 
	corrected bug in lm.c: pred. variance, depending on block size
	placed '#' on start of the 8 header lines of semivariograms, 
	to make life easier for gnuplot support
aug 94, gstat 1.2h
	added 'par:fit=1,gnuplot="file.gp"; functionality: see sem.c, 
	sem_fit.c this is for wls fitting of sill-nugget, and variogram 
	plotting using gnuplot. (added # in variogram header!)
aug 94, gstat 1.3 
	Added mivque.[ch], for mi[nv]que univariate covariance fitting
aug 94, gstat 1.3a
	added global kriging in [s]krige.c. Up to 80 times faster. Wow.
aug 94, gstat 1.3b
	added `area: "discr.dat",x=1,y=2', user defined area discretising
sep 94, gstat 1.3c
	fixed bug in CovSem_AB(): remember both within-block cov. and sem.
oct 94, gstat 1.3d
	added menu; corrected CheckX() for points();, but see 1.4b...
	added noaverage/average instead of "average=1"
	added "square" to use square neigborhood selections (no circle)
	changed variogram structure, now functions as part of structure.
	removed bug in getest, functions for MVS/MED, added par:quantile=..
dec 94, gstat 1.3e 
	menu evolved... added '#','?','-','&' options (try ?);
	added recalculation only at options Data, Calculate, cUtoff, &

----------gstat, 1995:
jan 94, gstat 1.4  
	menu.c elm-like, major rewrite. No hidden options, help added.  
jan 94, gstat 1.4a 
	corrected compile with STRICT_KR2 (no HAVE_CURSES)
	menu faster (incomplete screen rewrite after each key pressed)
	env. variable GNUPROLOG may define a prolog for gnuplot
jan 94, gstat 1.4b 
	corrected bug in select.c (cs with no range defined)
	env. variable GNUPLOT points gnuplot out, GSTATFMT "%g" variogram
	menu enhancements (echo Yes/No, Show)
	Yet another nasty bug in CheckX()!
feb 94, gstat  1.4c 
	header of variogram cloud corrected. Found bug in mopen.c (csf-lib)
	IsEas(): now use rewind() instead of fopen()+fclose().
	partial fitting: 0.2 @Nug() + 1 Sph(30) will now only fit Sph part.
	(removed obsolete Wls_fit_Sill())
	-L : allow local residuals.
	added dX to DATA: max X-distance allowed for semivar. calculation
	(removed obsolete `factor' from DATA. This can now be done
	in a more general fashion with X and dX).
	menu options: 'J': jgraph file; `&' jgraph-eps; '!': shell command.
Mar 94, gstat 1.4d 
	cleaned up random.c, removed rn.[ch] and Ran[0-3]()
	corrected estimation of sample variance (lm->s2) for weighted lm.
 	EnterData() (menu.c) now does not return without valid data read.
	temporary file names now come from tmpnam() or tempnam() [on UNIX];
	added MODE mode: more flexibility now (stratify whenever suitable)
	Fixed bug in Xvalid(): where.X, NA's on uk/lm for secondary vars
	Fixed bug (weights) in CalcLM(); added 's', and '+/-' in menu.
	Added DATE in date.h, changed program heading: -V now includes date.
	Now mean variances (V=..) are subtracted in variogram calculation.
Apr 94, gstat 1.4.e
	Made lm->SSR the corrected SS for mean, with p-1 df (Christensen p. 89)
	Added more.c (callable from menu). (removed later on...)
	syntax change: `pars:x=value;' to `set x=value;' (like gnuplot)
	Made `width' and `cutoff' global variables (`set cutoff=15,width=1').
	fit.c: now use lm.c routines for the fitting (estimation),
	Changed convergence criterion in WlsFit(), Mivque():

gstat/libcsf/src/setmv.c
/*
 * setmv.c
$Log: setmv.c,v $
Revision 2.0  1996/05/23 13:16:26  cees
csf2clean

Revision 1.1  1996/05/23 13:11:49  cees
Initial revision

Revision 1.3  1995/11/01 17:23:03  cees
.

 * Revision 1.2  1994/09/06  13:04:27  cees
 * changed for appCR field
 * added c2man docs
 *
 * Revision 1.1  1994/08/26  13:33:23  cees
 * Initial revision
 *
 */
#ifndef lint
 static const char *rcs_id = 
 "$Header: /home/pcrtree/SRC.RCS/libs/csf/RCS/setmv.c,v 2.0 1996/05/23 13:16:26 cees Exp $";
#endif


#include "csf.h"
#include "csfimpl.h"

/* set a memory location to a missing value
 * SetMV sets a memory location to a missing value
 * (using the application cell representation).
 * SetMV is quite slow but handy as in the example
 * below. In general one should use assignment for
 * integers (e.g. v = MV_UINT1) or the macro's
 * SET_MV_REAL4 and SET_MV_REAL8
 *
 * EXAMPLE
 * .so examples/border.tr
 */
void SetMV(
	const MAP *m, /* map handle */
	void *c)      /* write-only. location set to missing value */
{
	SetMVcellRepr(m->appCR, c);
}

/* set a memory location to a missing value
 * SetMVcellRepr sets a memory location to a missing value
 * (using the application cell representation).
 * In general one should use assignment for
 * integers (e.g. v = MV_UINT1) or the macro's
 * SET_MV_REAL4 and SET_MV_REAL8
 *
 */
void SetMVcellRepr(
	CSF_CR cellRepr, /* cell representation, one of the CR_* constants */
	void *c)      /* write-only. location set to missing value */
{
	switch (cellRepr)
	{
		case CR_INT1  : *((INT1 *)c) = MV_INT1;
				break;
		case CR_INT2  : *((INT2 *)c) = MV_INT2;
				break;
		case CR_INT4  : *((INT4 *)c) = MV_INT4;
				break;
		case CR_UINT1 : *((UINT1 *)c) = MV_UINT1;
				break;
		case CR_UINT2 : *((UINT2 *)c) = MV_UINT2;
				break;
		case CR_REAL8 :
				((UINT4 *)c)[1] = MV_UINT4;
		default       : POSTCOND(
					cellRepr == CR_REAL8 ||
					cellRepr == CR_REAL4 ||
					cellRepr == CR_UINT4 );
				*((UINT4 *)c) = MV_UINT4;
	}
} 
